pkg/controller/nodeipam/ipam/multi_cidr_priority_queue.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipam
+
+import (
+	"math"
+
+	cidrset "k8s.io/kubernetes/pkg/controller/nodeipam/ipam/multicidrset"
+)
+
+// A PriorityQueue implementation based on https://pkg.go.dev/container/heap#example-package-PriorityQueue
+
+// An PriorityQueueItem is something we manage in a priority queue.
+type PriorityQueueItem struct {
+	clusterCIDR *cidrset.ClusterCIDR
+	// labelMatchCnt is the first determinant of priority.
+	labelMatchCnt int
+	// label is a string representation of the labelSelector associated with the cidrSet.
+	label string
+	// index is needed by update and is maintained by the heap.Interface methods.
+	index int // The index of the item in the heap.
+}
+
+// A PriorityQueue implements heap.Interface and holds PriorityQueueItems.
+type PriorityQueue []*PriorityQueueItem
+
+func (pq PriorityQueue) Len() int { return len(pq) }
+
+func (pq PriorityQueue) Less(i, j int) bool {
+	if pq[i].labelMatchCnt != pq[j].labelMatchCnt {
+		// P0: CidrSet with higher number of matching labels has the highest priority.
+		return pq[i].labelMatchCnt > pq[j].labelMatchCnt
+	}
+
+	// If the count of matching labels is equal, compare the max allocatable pod CIDRs.
+	if pq[i].maxAllocatable() != pq[j].maxAllocatable() {
+		// P1: CidrSet with fewer allocatable pod CIDRs has higher priority.
+		return pq[i].maxAllocatable() < pq[j].maxAllocatable()
+	}
+
+	// If the value of allocatable pod CIDRs is equal, compare the node mask size.
+	if pq[i].nodeMaskSize() != pq[j].nodeMaskSize() {
+		// P2: CidrSet with a PerNodeMaskSize having fewer IPs has higher priority.
+		// For example, `27` (32 IPs) picked before `25` (128 IPs).
+		return pq[i].nodeMaskSize() > pq[j].nodeMaskSize()
+	}
+
+	// If the per node mask size are equal compare the CIDR labels.
+	if pq[i].label != pq[j].label {
+		// P3: CidrSet having label with lower alphanumeric value has higher priority.
+		return pq[i].label < pq[j].label
+	}
+
+	// P4: CidrSet having a smaller IP address value has a higher priority.
+	return pq[i].cidrLabel() < pq[j].cidrLabel()
+}
+
+func (pq PriorityQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *PriorityQueue) Push(x interface{}) {
```

Comment:
Do we need to take in interface{} or just PriorityQueueItem

Reply:
We need to implement the heap interface here, so we need to take in interface{}